### DISCUSSION TITLE
Objectives Evaluation

### DISCUSSION THREAD

**Discussion by meganindya** [sugarlabs/musicblocks-v4#3] [Ideas]
# Client

## Palette

- narrower and collapsible
- general (basic) blocks and niche (advanced) blocks
- group similar blocks
- plugin buttons

## Main Toolbar

- _run_ (full-speed) and _stop_; _run-slowly_ and _run-step-by-step_ to be in dropdown of _run_
- _new project_, _load project_, _export to sheet music_ to be grouped together
- canvas features: _save artwork_, _save animation (GIF)_, _record canvas (MPEG4)_, _record audio (MP3/WAV/OGG)_ to be grouped together
- _undo_-_redo_
- language
- guide

## State Toolbar

- global settings
- _statistics_

## Canvas

- `debatable` page mode (no infinite canvas)
- linear block stacks arrangement (no random position) with a canvas to the side

## Canvas Toolbar

- _maps_ (e.g. cartesian planes)
- _reset_
- rearrange blocks, toggle canvas mode (linear or free) to be grouped together
- _zoom_, _collapse_, _expand_, _snap-to-grid_ to be grouped together

## Functions

- history management (blocks and canvas)
- context management (e.g. `CanvasContext`, `AudioContext`, `MediaContext`, etc.)
- manage widgets based on feature API

# Broker

## Pool (Shared Memory)

- block tree

## Functions

- signal dispatcher (for client-engine) communication
- block framework
- serialize/deserialize block tree
- feature API middleware

# Engine

## Core

- single stack-based state management
- effective low-level transitions (e.g. directly work on pitch frequencies)
- feature API (e.g. _ADSR_ properties, time-measurement, etc.)

## Scheduler

- variable-based _Gantt-chart_ using `Tone.Transport`

## Utilities

- signal dispatcher (to client via broker)
- sheet music _export_ algorithms

**Comment by meganindya** [comment]
@walterbender did I miss anything? This one's sort of an exhaustive list of all features and their arrangement. I have been thinking about the organization throughout the week, and this makes most sense to me. The classes will should be fairly simple to identify if we get this module-based requirements right.

**Comment by walterbender** [comment]
Looks great. I am just uncertain about where the export belongs. We prob. want to keep a log of the notes played on the client side and then use a utility to convert that log into sheet music, etc. What do you think?

**Comment by meganindya** [comment]
Tbh I don't like the dependence of the sheet music conversion on the program running. How about we identify the sheet music just from the blocks itself? We can, in fact, do it on the fly. I understand there are interactive elements but can't we get around that?

**Comment by walterbender** [comment]
Let's think about it some more. I think losing interactivity is a major step backwards. For example, we had a use case in class this morning: The kids are writing games... there is a theme music sound track and differ music that comes in during different events --- approaching a monster, defeating a monster, etc. The kids what to print the sheet music as a transcript of their game play. I think it is a pretty cool idea that we would not want to lose.

**Comment by meganindya** [comment]
I personally don't like the experience where the program plays without any effect, i.e. no sound and artwork, during the translation. Perhaps, a dynamic way 
(when program runs) and a static way (from the block tree)? I'd very much want a visual timeline, but I can't really think of a non-static way.

**Comment by walterbender** [comment]
The way I had originally done the exports (and I don't recall why we changed it) was that you exported whatever you last ran, so it was not a matter of rerunning the code without any effect. Maybe we go back to that?

**Comment by walterbender** [comment]
Or maybe: start recording now; end recording; export.

**Comment by pikurasa** [comment]
Do we want to add any things related to collaboration and "version control" to this list? (Or is that for a separate list, for the server?)
